File: client/src/presence.ts
export type PresenceState =
  | "idle"
  | "working"
  | "using_computer"
  | "waiting_for_approval"
  | "waiting_for_takeover"
  | "collaborating"
  | "failed";

export type AgentPresence = {
  state: PresenceState | string;
  label: string;
  detail?: string;
};

export type PresenceRun = {
  bot_id?: string;
  status: string;
  error?: string;
};

export type PresenceHandoff = {
  source_bot_id: string;
  target_bot_id: string;
  status?: string;
};

export type PresenceComputer = {
  running: boolean;
  takeover: boolean;
  agent_control?: boolean;
};

export function deriveAgentPresence(input: {
  botID: string;
  run?: PresenceRun | null;
  computer?: PresenceComputer | null;
  handoffs?: PresenceHandoff[];
}): AgentPresence {
  const run = input.run;
  const computer = input.computer;
  const active = Boolean(run && runActive(run.status));
  if (run?.status === "waiting_for_approval") {
    return {
      state: "waiting_for_approval",
      label: "Needs approval",
      detail: "Waiting for an explicit Allow or Deny",
    };
  }
  if (active && computer?.takeover) {
    return {
      state: "waiting_for_takeover",
      label: "Needs takeover",
      detail: "Human control of the Agent Computer",
    };
  }
  if (active && computer?.running && (computer.agent_control || computer.takeover)) {
    return {
      state: "using_computer",
      label: "Using computer",
      detail: "Browser or desktop work on the isolated computer",
    };
  }
  if (collaborating(input.botID, input.handoffs ?? [])) {
    return {
      state: "collaborating",
      label: "Collaborating",
      detail: active ? "Working with a teammate" : "Asked another Agent",
    };
  }
  if (run && (run.status === "queued" || run.status === "running")) {
    return {
      state: "working",
      label: run.status === "queued" ? "Queued" : "Working",
      detail: `Run ${run.status}`,
    };
  }
  if (run && (run.status === "failed" || run.status === "blocked")) {
    return {
      state: "failed",
      label: "Failed",
      detail: run.error || `Last run ${run.status}`,
    };
  }
  return { state: "idle", label: "Idle" };
}

function runActive(status: string) {
  return status === "queued" || status === "running" || status === "waiting_for_approval";
}

function collaborating(botID: string, handoffs: PresenceHandoff[]) {
  return handoffs.some((handoff) => {
    const status = handoff.status ?? "";
    if (status !== "queued" && status !== "running" && status !== "waiting") {
      return false;
    }
    return handoff.source_bot_id === botID || handoff.target_bot_id === botID;
  });
}

export function engineShortLabel(harness?: string) {
  switch (harness) {
    case "grok":
    case "grok_build":
      return "Grok";
    case "codex_app_server":
      return "Codex";
    case "opencode":
      return "OpenCode";
    case "pi":
      return "Pi";
    default:
      return "";
  }
}

File: internal/domain/presence.go
package domain

const (
	PresenceIdle              = "idle"
	PresenceWorking           = "working"
	PresenceUsingComputer     = "using_computer"
	PresenceWaitingApproval   = "waiting_for_approval"
	PresenceWaitingTakeover   = "waiting_for_takeover"
	PresenceCollaborating     = "collaborating"
	PresenceFailed            = "failed"
)

// AgentPresence is a computed, non-durable roster state. It is never stored.
type AgentPresence struct {
	State  string `json:"state"`
	Label  string `json:"label"`
	Detail string `json:"detail,omitempty"`
}

// ComputerPresence is the workspace computer snapshot used to derive roster
// state. The computer is shared; only an Agent with an active run is shown as
// using it.
type ComputerPresence struct {
	Running      bool
	AgentControl bool
	Takeover     bool
}

// PresenceInput is the exact evidence DeriveAgentPresence may consider.
type PresenceInput struct {
	BotID    string
	Run      *Run
	Computer ComputerPresence
	Handoffs []Handoff
}

// DeriveAgentPresence maps existing run, computer, and handoff evidence onto
// one user-visible roster state. Higher-urgency states win.
func DeriveAgentPresence(input PresenceInput) AgentPresence {
	run := input.Run
	active := run != nil && runActive(run.Status)
	if run != nil && run.Status == "waiting_for_approval" {
		return AgentPresence{State: PresenceWaitingApproval, Label: "Needs approval", Detail: "Waiting for an explicit Allow or Deny"}
	}
	if active && input.Computer.Takeover {
		return AgentPresence{State: PresenceWaitingTakeover, Label: "Needs takeover", Detail: "Human control of the Agent Computer"}
	}
	if active && input.Computer.Running && (input.Computer.AgentControl || input.Computer.Takeover) {
		return AgentPresence{State: PresenceUsingComputer, Label: "Using computer", Detail: "Browser or desktop work on the isolated computer"}
	}
	if collaborating(input.BotID, input.Handoffs) {
		detail := "Asked another Agent"
		if run != nil && runActive(run.Status) {
			detail = "Working with a teammate"
		}
		return AgentPresence{State: PresenceCollaborating, Label: "Collaborating", Detail: detail}
	}
	if run != nil && (run.Status == "queued" || run.Status == "running") {
		label := "Working"
		if run.Status == "queued" {
			label = "Queued"
		}
		return AgentPresence{State: PresenceWorking, Label: label, Detail: "Run " + run.Status}
	}
	if run != nil && (run.Status == "failed" || run.Status == "blocked") {
		detail := run.Error
		if detail == "" {
			detail = "Last run " + run.Status
		}
		return AgentPresence{State: PresenceFailed, Label: "Failed", Detail: detail}
	}
	return AgentPresence{State: PresenceIdle, Label: "Idle"}
}

func runActive(status string) bool {
	switch status {
	case "queued", "running", "waiting_for_approval":
		return true
	default:
		return false
	}
}

func collaborating(botID string, handoffs []Handoff) bool {
	if botID == "" {
		return false
	}
	for _, handoff := range handoffs {
		switch handoff.Status {
		case HandoffStatusQueued, HandoffStatusRunning, HandoffStatusWaiting:
		default:
			continue
		}
		if handoff.SourceBotID == botID || handoff.TargetBotID == botID {
			return true
		}
	}
	return false
}

func PresenceLabel(state string) string {
	switch state {
	case PresenceWaitingApproval:
		return "Needs approval"
	case PresenceWaitingTakeover:
		return "Needs takeover"
	case PresenceUsingComputer:
		return "Using computer"
	case PresenceCollaborating:
		return "Collaborating"
	case PresenceWorking:
		return "Working"
	case PresenceFailed:
		return "Failed"
	default:
		return "Idle"
	}
}
